src/lua/loaders.lua

```diff
@@ -58,36 +58,63 @@ local function traverse_path(path)
     return paths
 end
 
--- Generate a search function, which performs searching through
--- templates setup in options.
+-- Generate a path builder function, which yields the set of the
+-- paths to be searched through for the desired module.
 --
--- @param path_fn function which returns a base path for the
---     resulting template
+-- @param basepath_fn function which returns a base path for the
+--     resulting path set
 -- @param templates table with lua search templates
 -- @param need_traverse bool flag which tells search function to
 --     build multiple paths by expanding base path up to the
 --     root ('/')
--- @return a searcher function which builds a path template and
---     calls package.searchpath
-local function gen_search_func(path_fn, templates, need_traverse)
-    assert(type(path_fn) == 'function', 'path_fn must be a function')
+-- @return a path builder function which builds and yields the
+--     resulting path set
+local function gen_path_builder(basepath_fn, templates, need_traverse)
+    assert(type(basepath_fn) == 'function', 'basepath_fn must be a function')
     assert(type(templates) == 'table', 'templates must be a table')
 
-    return function(name)
-        local path = path_fn() or '.'
-        local paths = need_traverse and traverse_path(path) or { path }
+    return function()
+        local base = basepath_fn() or '.'
+        local dirs = need_traverse and traverse_path(base) or { base }
 
         local searchpaths = {}
 
-        for _, path in ipairs(paths) do
+        for _, dir in ipairs(dirs) do
             for _, template in pairs(templates) do
-                table.insert(searchpaths, minifio.pathjoin(path, template))
+                table.insert(searchpaths, minifio.pathjoin(dir, template))
             end
         end
 
-        local searchpath = table.concat(searchpaths, ';')
+        return table.concat(searchpaths, ';')
+    end
+end
+
+local path = {
+    package = function() return package.path end,
+    cwd = {
+        dot = gen_path_builder(searchroot, LUA_TEMPLATES),
+        rocks = gen_path_builder(searchroot, ROCKS_LUA_TEMPLATES, true),
+    }
+}
+
+local cpath = {
+    package = function() return package.cpath end,
+    cwd = {
+        dot = gen_path_builder(searchroot, LIB_TEMPLATES),
+        rocks = gen_path_builder(searchroot, ROCKS_LIB_TEMPLATES, true),
+    }
+}
```

Comment:
On the first glance I read it as `path` and `cpath` objects and only then understood that it is a way to group the data.

I would also highlight that we have:

* Module level variables with quite common names `path` and `cpath` that contain functions, not paths.
* Functions of the same kind on different nesting levels: `path.package` and `path.cwd.dot`.

The alternative would be, for example, the following.

```lua
local path_builders = {
    ['path.package'] = <...>,
    ['path.cwd.dot'] = <...>,
    ['path.cwd.rocks'] = <...>,
    <...>
}
```

Feel free to ignore. This is a matter of taste.

BTW, `{path,cpath}.package` goes after `{path,cpath}.cwd.{dot,rocks}` in the searchers/loaders order. It would be good to keep this order in these definitions too.

Reply:
I agree this is the matter of taste, so I prefer to left this intact for now. Ignoring.